pkg/api/query/grpc.go

```diff
@@ -101,24 +102,36 @@ func (g *GRPCAPI) Query(request *querypb.QueryRequest, server querypb.Query_Quer
 		query.NoopSeriesStatsReporter,
 	)
 
-	var engine promql.QueryEngine
 	engineParam := request.Engine
 	if engineParam == querypb.EngineType_default {
 		engineParam = g.defaultEngine
 	}
 
+	var qry promql.Query
 	switch engineParam {
 	case querypb.EngineType_prometheus:
-		engine = g.engineFactory.GetPrometheusEngine()
+		queryEngine := g.engineFactory.GetPrometheusEngine()
+		qry, err = queryEngine.NewInstantQuery(ctx, queryable, promql.NewPrometheusQueryOpts(false, lookbackDelta), request.Query, ts)
+		if err != nil {
+			return err
+		}
 	case querypb.EngineType_thanos:
-		engine = g.engineFactory.GetThanosEngine()
+		queryEngine := g.engineFactory.GetThanosEngine()
+		plan, err := logicalplan.Unmarshal(request.QueryPlan.GetJson())
+		if err != nil {
+			// fallback to using query string if plan is not provided.
+			qry, err = queryEngine.NewInstantQuery(ctx, queryable, promql.NewPrometheusQueryOpts(false, lookbackDelta), request.Query, ts)
```

Comment:
There is indeed a bug here, since we would always call `queryEngine.NewInstantQueryFromPlan`. Maybe we should move this logic into a function so that we can use early returns.

Reply:
Yep, realized after push. Good idea about refactoring to method. Added a simple unit test as well.